python/cudf/cudf/core/abc.py

```diff
@@ -0,0 +1,56 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.
+
+import abc
+import pickle
+from abc import abstractmethod
+
+import rmm
+
+
+class Serializable(abc.ABC):
+    @abstractmethod
+    def serialize(self):
+        pass
+
+    @classmethod
+    @abstractmethod
+    def deserialize(cls, header, frames):
+        pass
+
+    def device_serialize(self):
+        header, frames = self.serialize()
+        assert all((type(f) is cudf.core.buffer.Buffer) for f in frames)
+        header["type-serialized"] = pickle.dumps(type(self))
+        header["lengths"] = [f.nbytes for f in frames]
+        return header, frames
+
+    @classmethod
+    def device_deserialize(cls, header, frames):
+        for f in frames:
```

Comment:
This isn't going to be true with pack/unpack serialization which packs into a single host buffer that stores metadata, and a single device buffer that stores data.

Reply:
Yep, that makes sense. Happy to change this as we see fit.

The bigger idea here is that when we make these changes we can now do them in one place. So hopefully that makes pack/unpack and other changes in the future easier :)

Reply:
Agreed. I think then we can leave this as is for now. :)

Reply:
Added PR ( https://github.com/rapidsai/cudf/pull/5309 ), which should handle a mixture of host and device frames.